src/sqs/index.ts

```diff
@@ -258,23 +311,67 @@ export class SqsQueue<MessagePayload> {
 			executionId: uuidv4(),
 			queue: this.queueName})
 
-		const context: Context<MessagePayload> = {message, payload, log }
+		const maybeReceivedCount: MessageAttributeValue | undefined = message.MessageAttributes ? message.MessageAttributes["ApproximateReceiveCount"] : undefined;
+
+		const receiveCount = maybeReceivedCount ? parseInt(maybeReceivedCount?.StringValue || "1") : 1
 
-		log.info("Sqs message received");
+		const context: Context<MessagePayload> = {message, payload, log, receiveCount: receiveCount}
+
+		log.info(`Sqs message received. Receive count: ${receiveCount}`);
 
 		try {
 			const messageProcessingStartTime = new Date().getTime();
-			await this.messageHandler.handle(context)
+
+			// Change message visibility timeout to the max processing time
+			// plus some room for error handling in case of timeout
+			await this.changeVisabilityTimeout(message, this.timeoutSec + 2, log);
+
+			const timeoutPromise = new Promise((_, reject) =>
+				setTimeout(() => reject(new SqsTimeoutError()), this.timeoutSec*1000)
+			);
+
+			await Promise.race([this.messageHandler.handle(context), timeoutPromise])
+
 			const messageProcessingDuration = new Date().getTime() - messageProcessingStartTime;
 			this.sendProcessedMetrics(messageProcessingDuration);
 			await this.deleteMessage(message, log)
 		} catch (err) {
+
+			const errorHandlingResult = this.errorHandler.handle(err, context);
+			if(!errorHandlingResult.retryable ||
+				(errorHandlingResult.skipDlq && context.receiveCount >= this.maxAttempts)) {
+				await this.deleteMessage(message, log)
+			} else {
+				if(errorHandlingResult.retryDelaySec) {
+					await this.changeVisabilityTimeout(message, errorHandlingResult.retryDelaySec, log);
+				}
+			}
+
 			//TODO Add error handling
 			statsd.increment(sqsQueueMetrics.failed, this.metricsTags)
 			log.error({err}, "error executing sqs message")
 		}
 	}
 
+	private async changeVisabilityTimeout(message: Message, timeout: number, logger: Logger): Promise<void> {
+
+		if(!message.ReceiptHandle) {
+			logger.error("NO Receipt Handle on a message")
+			return;
+		}
+
+		const params : ChangeMessageVisibilityRequest = {
+			QueueUrl: this.queueUrl,
+			ReceiptHandle: message.ReceiptHandle || "",
```

Comment:
don't need the `|| ""`

Reply:
I am getting typescript errors otherwise hence message.ReceiptHandle is optional. The "" value will never be here because in the lines above we check if it is empty and return. 

Would would you suggest to convert optional to non-optional?

Reply:
I'm fairly certain that if you have a check with a return beforehand, TS assumes the type is now specified and not optional and should work.  I just tried this exact code without the || and it worked fine.